server/sonar-server/src/main/java/org/sonar/server/computation/step/QualityGateEventsStep.java

```diff
@@ -0,0 +1,133 @@
+import org.sonar.server.computation.event.EventRepository;
+import org.sonar.server.computation.measure.MeasureRepository;
+
+public class QualityGateEventsStep implements ComputationStep {
+  private final TreeRootHolder treeRootHolder;
+  private final EventRepository eventRepository;
+  private final MeasureRepository measureRepository;
+
+  public QualityGateEventsStep(TreeRootHolder treeRootHolder, EventRepository eventRepository, MeasureRepository measureRepository) {
+    this.eventRepository = eventRepository;
+    this.measureRepository = measureRepository;
+    this.treeRootHolder = treeRootHolder;
+  }
+
+  @Override
+  public void execute() {
+    new DepthTraversalTypeAwareVisitor(Component.Type.PROJECT, DepthTraversalTypeAwareVisitor.Order.PRE_ORDER) {
+      @Override
+      public void visitProject(Component project) {
+        executeForProject(project);
+      }
+    }.visit(treeRootHolder.getRoot());
+  }
+
+  private void executeForProject(Component project) {
+    Optional<BatchReport.Measure> currentStatus = measureRepository.findCurrent(project, CoreMetrics.ALERT_STATUS);
+    if (!currentStatus.isPresent()) {
+      return;
+    }
+
+    Optional<MeasureDto> previousAlertStatus = measureRepository.findPrevious(project, CoreMetrics.ALERT_STATUS);
+
+    checkQualityGateStatusChange(project, currentStatus.get(), previousAlertStatus);
+  }
+
+  private void checkQualityGateStatusChange(Component project, BatchReport.Measure currentStatus, Optional<MeasureDto> pastStatus) {
+    String alertText = currentStatus.getAlertText();
+    MetricLevel alertLevel = MetricLevel.valueOf(currentStatus.getAlertStatus());
+    if (pastStatus.isPresent()) {
+      MetricLevel pastAlertLevel = MetricLevel.valueOf(pastStatus.get().getAlertStatus());
+      if (pastAlertLevel != alertLevel) {
+        // The alert status has changed
+        String alertName = buildName(pastStatus.get(), currentStatus);
+        createEvent(project, alertName, alertText);
+        // There was already a Orange/Red alert, so this is no new alert: it has just changed
+        // boolean isNewAlert = pastAlertLevel == MetricLevel.OK;
+        // notifyUsers(project, alertName, alertText, alertLevel, isNewAlert);
+      }
+    } else if (alertLevel != MetricLevel.OK) {
+      // There were no defined alerts before, so this one is a new one
+      String alertName = buildName(currentStatus);
+      createEvent(project, alertName, alertText);
+      // notifyUsers(project, alertName, alertText, alertLevel, true);
+    }
+  }
+
+  private enum MetricLevel {
+    OK("Green"), WARN("Orange"), ERROR("Red");
+
+    private String colorName;
+
+    MetricLevel(String colorName) {
+      this.colorName = colorName;
+    }
+
+    public String getColorName() {
+      return colorName;
+    }
+  }
+
+  // protected void notifyUsers(Component project, String alertName, String alertText, MetricLevel alertLevel, boolean isNewAlert) {
```

Comment:
I forsee it will be done in the same step, hence the code left commented out for now